hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.SqlRowMetadata;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.SqlRowImpl;
+import com.hazelcast.sql.impl.row.HeapRow;
+import com.hazelcast.sql.impl.row.Row;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * Client-side cursor.
+ */
+public class SqlClientResult implements SqlResult {
+
+    private final SqlClientService service;
+    private final Connection connection;
+    private final QueryId queryId;
+    private final SqlRowMetadata rowMetadata;
+    private final ClientIterator iterator = new ClientIterator();
+
+    private int cursorBufferSize;
+    private boolean closed;
+    private boolean iteratorAccessed;
+
+    public SqlClientResult(
+        SqlClientService service,
+        Connection connection,
+        QueryId queryId,
+        SqlRowMetadata rowMetadata,
+        SqlPage page,
+        int cursorBufferSize
+    ) {
+        this.service = service;
+        this.connection = connection;
+        this.queryId = queryId;
+        this.rowMetadata = rowMetadata;
+        this.cursorBufferSize = cursorBufferSize;
+
+        iterator.onNextPage(page);
+    }
+
+    @Nonnull
+    @Override
+    public SqlRowMetadata getRowMetadata() {
+        return rowMetadata;
+    }
+
+    @Override
+    @Nonnull
+    public Iterator<SqlRow> iterator() {
+        if (!iteratorAccessed) {
+            iteratorAccessed = true;
+
+            return iterator;
+        } else {
+            throw new IllegalStateException("Iterator could be requested only once");
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            if (!closed) {
+                if (iterator.last) {
+                    // The last page accessed, so the remote cursor is already closed. No-op.
+                    return;
+                }
+
+                service.close(connection, queryId);
+            }
+        } finally {
+            closed = true;
```

Comment:
If an exception is thrown from` service.close(connection, queryId)` then we still mark the iterator as closed while it is not?

Reply:
The only reason for the exception here is a broken connection. If it is broken, there is no sense trying to keep the result opened, because no more rows could be fetched anyway.